src/qibolab/identifier.py

```diff
@@ -0,0 +1,63 @@
+from enum import Enum
+from typing import Annotated, Optional, Union
+
+from pydantic import Field, TypeAdapter, model_serializer, model_validator
+
+from .serialize import Model
+
+QubitId = Annotated[Union[int, str], Field(union_mode="left_to_right")]
+"""Type for qubit names."""
+
+
+# TODO: replace with StrEnum, once py3.10 will be abandoned
+# at which point, it will also be possible to replace values with auto()
+class ChannelType(str, Enum):
+    """Names of channels that belong to a qubit.
+
+    Not all channels are required to operate a qubit.
+    """
+
+    PROBE = "probe"
+    ACQUISITION = "acquisition"
+    DRIVE = "drive"
+    DRIVE12 = "drive12"
+    DRIVE_CROSS = "drive_cross"
+    FLUX = "flux"
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+
+class ChannelId(Model):
+    """Unique identifier for a channel."""
+
+    qubit: QubitId
```

Comment:
> Isn't this a bit cyclic? We have, `Qubit` -> `Channel` -> `ChannelId` -> `QubitId`. Of course it is not cyclic in terms of objects because `QubitId != Qubit`, but at least conceptually wise it kind of is.

The problem of your picture is `Channel` -> `ChannelId`: an object should not be aware of its own identifier, because that's how the other ones are calling you, but you don't really need to call yourself.

> On the other hand, I am not sure if making channels depend on qubits (in this case `QubitId`) is a good idea. I understand here you are trying to standardize channel names. However, in general, channels are a more primitive idea than qubits, as in qubits need channels to be controlled, not the other way around. Channels could survive without qubits, potentially in non-quantum computing applications.

True, but Qibolab is mostly about quantum computing.

In any case, the main reason why `ChannelId` now contains `QubitId` is exactly because `Qubit` contains the `Channel`: if you have a pulse sequence, you truly want to be able to get a handle on the channel, given a `platform`. So, the `ChannelId` is a practical path to retrieve the `Channel` you need.
If instead we give up on `Qubit` as a `Channel` collection (and thus on the `Qubit` object tout-court, because that's what is left), we don't need to navigate, but just grab the channel from the channels' container (a single platform attribute).

Reply:
> The stricter would be `dict[ChannelId, Channel]`, however that may be less user friendly when creating sequences, as
> ```py
> sequence.append((platform.channels["0/drive"], Pulse(...)))
> ```
> is simpler than
> ```py
> sequence.append((platform.channels[ChannelId(...)], Pulse(...)))
> ```
> even though we could also alleviate this by defining an appropriate getter in the `Platform`.

Here you're a bit unfair, since the current layout is just:
```py
sequence.append((ChannelId(...), Pulse(...)))
```

And that's better than everything: a pulse sequence should not be bound to a specific `platform`, but just a sequence of operations. If you change platform, but you keep the same channel IDs, the instructions should still execute (which is relevant for serialization, when you dump and load the sequence again).

In any case, without the `Qubit`, we could definitely also have a simpler `ChannelId`, just reminiscent of the role of the channel (i.e. the `ChannelType`)[*], while retrieving the qubit association from somewhere else.

In a sense, this is really a leftover: by now, everything is working by named reference, without nesting objects (other than in `parameters.json`), and that was the idea of the components configs. But we could treat the actual components in the same way, and just retain the name in the `Qubit`s (to answer the question: which is the channel for this qubit's flux?)

[*]: we could drop even the `ChannelType`, but this information is instead relevant even in the `PulseSequence`, so I'd try to keep it somehow - the other option is to make `PulseSequence` a full-blown object, not just a list, and register the channel's roles in there (not sure about which could be an effective interface)

Reply:
> The problem of your picture is `Channel` -> `ChannelId`: an object should not be aware of its own identifier, because that's how the other ones are calling you, but you don't really need to call yourself.

But isn't the `Channel` aware `ChannelId` in this PR? `Channel.name` is the `ChannelId`. The -> in my original message basically means "has a reference to". `Qubit` has references to `Channel`s, `Channel` has reference to its `ChannelId`, etc..

> In any case, the main reason why `ChannelId` now contains `QubitId` is exactly because `Qubit` contains the `Channel`: if you have a pulse sequence, you truly want to be able to get a handle on the channel, given a `platform`. So, the `ChannelId` is a practical path to retrieve the `Channel` you need. 

I think I don't understand this. How `ChannelId` helps retrieving `Channel`s? I think the way to retrieve them as it is now is still `platform.qubits[qubit_name].drive`, etc. but that was working even before this PR.

> If instead we give up on `Qubit` as a `Channel` collection (and thus on the `Qubit` object tout-court, because that's what is left), we don't need to navigate, but just grab the channel from the channels' container (a single platform attribute).

Yes, that's basically my proposal.

Reply:
> But isn't the `Channel` aware `ChannelId` in this PR? `Channel.name` is the `ChannelId`. The -> in my original message basically means "has a reference to". `Qubit` has references to `Channel`s, `Channel` has reference to its `ChannelId`, etc..

Yes, it was not a mistake in your reconstruction, just what I see as problematic.

> I think I don't understand this. How `ChannelId` helps retrieving `Channel`s? I think the way to retrieve them as it is now is still `platform.qubits[qubit_name].drive`, etc. but that was working even before this PR.

See below https://github.com/qiboteam/qibolab/pull/970#discussion_r1723521131

In principle, just reading the `PulseSequence` you didn't know it was the drive (because names were just whatever `str`). So, you had first to find (somehow) that your channel was referred to that qubit, and it was drive. Then you could access.
In practice, you were doing this by keeping a copy
https://github.com/qiboteam/qibolab/blob/b49bb420bb248acb1bfbbe14f2672cf6156a0a85/src/qibolab/instruments/qm/controller.py#L140
and never accessing the channels in the `Qubit` objects
https://github.com/qiboteam/qibolab/blob/b49bb420bb248acb1bfbbe14f2672cf6156a0a85/src/qibolab/platform/platform.py#L103

Reply:
In the previous comments I am mainly talking about how users (mostly qibocal) are supposed to access channels and create sequences, not the driver.

The driver certainly cannot access channels from `Qubit`s because it does not have access to `Qubit`s and it shouldn't. But the driver also does not care whether a channel is a drive of a particular qubit or something else. It only cares what are the pulses playing on that channel (provided by the sequence) and what is the config (frequencies, offsets, potentially LOs - provided by the `configs` dict). It may only need to distinguish acquisition channels from the rest, but I think even this is not necessary now since there is a separate `Acquisition` instruction.

Reply:
Correct. But the driver should also be the only one to care about the `Channel` object. So, it makes much more sense to locate the `Channel` in the instrument, rather than in the `Qubit` in the platform.